lib/internal/modules/esm/resolve.js

```diff
@@ -111,17 +111,17 @@ function emitLegacyIndexDeprecation(url, packageJSONUrl, base, main) {
   const path = fileURLToPath(url);
   const pkgPath = fileURLToPath(new URL('.', packageJSONUrl));
   const basePath = fileURLToPath(base);
-  if (main)
+  if (main && resolve(pkgPath, main) !== path)
     process.emitWarning(
-      `Package ${pkgPath} has a "main" field set to ${JSONStringify(main)}, ` +
+      `Package ${pkgPath} has a "main" field set to "${main}", ` +
       `excluding the full filename and extension to the resolved file at "${
         StringPrototypeSlice(path, pkgPath.length)}", imported from ${
         basePath}.\n Automatic extension resolution of the "main" field is ` +
       'deprecated for ES modules.',
       'DeprecationWarning',
       'DEP0151',
     );
-  else
+  else if (!main)
```

Comment:
since main can be a falsy value, can you update the jsdoc declaration and add a test?